### ISSUE TITLE
Older implementation of learning without forgetting performed better

### ISSUE THREAD

**Issue by tachyonicClock** [ContinualAI/avalanche#825]
🐛 **Describe the bug**
The old implementation of LwF performed much better than the current one. Is this working as intended or has a bug been introduced? As a further note what sort of testing would prevent this sort of regression?

🐜 **To Reproduce**
```python
from torch.optim import SGD
from torch.nn import CrossEntropyLoss
from avalanche.benchmarks.classic import SplitMNIST
from avalanche.evaluation.metrics import accuracy_metrics
from avalanche.models import SimpleMLP
from avalanche.logging import InteractiveLogger
from avalanche.training.plugins import EvaluationPlugin
from avalanche.training.strategies import LwF

scenario = SplitMNIST(n_experiences=5)
model = SimpleMLP(num_classes=scenario.n_classes)

eval_plugin = EvaluationPlugin(
    accuracy_metrics(minibatch=True, epoch=True, experience=True, stream=True),
    loggers=[InteractiveLogger()]
)

cl_strategy = LwF(model, SGD(model.parameters(), lr=0.001, momentum=0.9),
    CrossEntropyLoss(), train_mb_size=500, train_epochs=1, eval_mb_size=100, alpha=10, temperature=2.0,
    evaluator=eval_plugin)

print('Starting experiment...')
for i, experience in enumerate(scenario.train_stream):
    print("Start of experience: ", experience.current_experience)
    print("Current Classes: ", experience.classes_in_this_experience)

    res = cl_strategy.train(experience)
    print('Training completed')
    print('Computing accuracy on the whole test set')
    cl_strategy.eval(scenario.test_stream[:i+1])
```

**New implementation**
```
100%|██████████| 19/19 [00:00<00:00, 52.78it/s]
> Eval on experience 4 (Task 0) from test stream ended.
	Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp004 = 0.0021
-- >> End of eval phase << --
	Top1_Acc_Stream/eval_phase/test_stream/Task000 = 0.2298
```

**Old implementation as of 5356591e2355fbf2aa3d5c0dd5f7bc8613991cff**
```
100%|██████████| 21/21 [00:00<00:00, 47.12it/s]
> Eval on experience 4 (Task 0) from test stream ended.
	Top1_Acc_Exp/eval_phase/test_stream/Task000/Exp004 = 0.9178
-- >> End of eval phase << --
	Top1_Acc_Stream/eval_phase/test_stream = 0.4413
```

**Comment by AndreaCossu** [comment]
Hi @tachyonicClock . In the [reproducible continual learning repository](https://github.com/ContinualAI/reproducible-continual-learning) we tested LwF and we also encounter this issue. 
Old version uses distillatio on the entire output layer (thus assuming to know how many classes there will be in total). Current version distills only the previously active units. This reduces the performances on Split MNIST but it is a more realistic version. We decided to keep the last version of LwF since it is aligned with performances reported in other papers.

**Comment by tachyonicClock** [comment]
I see. Thanks for the clarification, I greatly appreciate it. I'm glad it's working as intended and we have the reproducible continual learning repository as a way to detect these sort of issues.

**Comment by tachyonicClock** [comment]
@AndreaCossu If you had to speculate, why do you think this reduces performance? Is it simply that it reduces the amount of "dark knowledge" captured by the distillation loss (I'm no expert on knowledge distillation).
Additionally I'm no longer convinced that having multiple classes breaks the setting. Surely the number of neurons on the output layer is largely unimportant. For example we could pick x neurons and add additional ones if needed and this would help with knowledge distillation. Alternatively the second to last layer could be used for distillation circumventing the criticism but still facilitating a large amount of knowledge to be captured.

The issue of reproducibility and matching other implementations is more significant. However shouldn't we explore the ways in which this bug could be exploited since it significantly improves results? Let me know if I'm mistaken as I said I'm not an expert on knowledge distillation.

**Comment by AntonioCarta** [comment]
If you think about it, the unused units will be penalized during training and will have high negative activations at the end of the experience (i.e. zero probability). As a result, at the next experience the knowledge distillation will push the new units (which are used at this step, but not the previous ones) towards negative activations (again, zero probability). However, since these new units are used now the model will also see positive examples.

Probably we can try to recover the old performance by 
- playing around with the hyperparameters.
- initializing the new units in a smarter way (?)
- adding an activation penalization on the new units.

> Surely the number of neurons on the output layer is largely unimportant. For example we could pick x neurons and add additional ones if needed and this would help with knowledge distillation. 

I think this is incorrect, because that's not how knowledge distillation and softmax  activations work. See above. Adding useless units does not have the same effect.

> The issue of reproducibility and matching other implementations is more significant. However shouldn't we explore the ways in which this bug could be exploited since it significantly improves results?

I agree with you. The fact that the old version performed much better is very interesting, but we should try to understand what it is that made it better (hyperparameters? init? penalization?). If you discover something useful let us know.

**Comment by tachyonicClock** [comment]
@AntonioCarta 
The point of soft-max is that it illuminates the slight differences in the output layer. Essentially the teacher network is saying "that potato looks like a potato but it also looks just a bit like a dog". The issue seems to be the fact that "dark knowledge" is lost when only a few activation units are used. A problem that is very significant in some of my research, because I am using a pre-trained resnet.

The issue that training (the traditional part) drives all the probabilities down is notable. Its true that before training on an experiences that experiences' activation are meaningless at least for a randomly initialised model. However I think it's clear empirically that the additional knowledge (or sort of regularisation) is significant (**especially if a pretrained model is used**). I concede that my original idea to add useless outputs is a little silly, but a revised version would be to use a different non-final layer close to the end to perform distillation with. I'm not suggesting we do that for LwF but it might be advantageous.
